include/rmm/mr/device/fixed_size_memory_resource.hpp

```diff
@@ -0,0 +1,235 @@
+
+  /**---------------------------------------------------------------------------*
+   * @brief Get free and available memory for memory resource
+   *
+   * @throws std::runtime_error if we could not get free / total memory
+   *
+   * @param stream the stream being executed on
+   * @return std::pair with available and free memory for resource
+   *---------------------------------------------------------------------------**/
+  std::pair<std::size_t, std::size_t> do_get_mem_info( cudaStream_t stream) const override {
+    return std::make_pair(0, 0);  
+  }
+
+  // return a block from the free list associated with the specified stream, and move all other 
+  // blocks in the list to the no-sync free list.
+  void* block_from_sync_list(cudaStream_t stream) {
+    free_list& blocks = sync_blocks_.at(stream);
+    void* p = nullptr;
+    if (blocks.size() > 0) {
+       p = blocks.front();
+       blocks.pop_front();
+    }
+    // insert all remaining blocks from sync list into no-sync list
+    if (p != nullptr) { // found one
+      cudaError_t result = cudaStreamSynchronize(stream);
+
+      if (result != cudaErrorInvalidResourceHandle && // stream deleted
+          result != cudaSuccess)                      // stream synced
+        throw std::runtime_error{"cudaStreamSynchronize failure"};
+
+      // insert all other blocks into the no_sync list
+      no_sync_blocks_.insert(no_sync_blocks_.end(), blocks.cbegin(), blocks.cend());
+      blocks.clear();
+
+      // remove this stream from the freelist
+      sync_blocks_.erase(stream);
+    }
+    return p;
+  }
+
+  // Returns a pointer to a free block. Attempts to return a block that is not associated with
+  // a stream first, and if there are none, returns a block from a sync free list. If no blocks 
+  // are available, allocates from the upstream heap resource
+  void* get_block(cudaStream_t stream) {
+    // Try to find a larger block that doesn't require syncing
+    void* p = nullptr;
+    if (no_sync_blocks_.size() > 0) {
+       p = no_sync_blocks_.front();
+       no_sync_blocks_.pop_front();
+    }
+
+    // nothing in no-sync free list, look for one on a stream
+    if (p == nullptr) {
+
+      // Try to find a larger block in a different stream
+      for (auto s : sync_blocks_) {
+        if (s.first != stream) p = block_from_sync_list(s.first);
+      }
+
+      // nothing available in other streams, look in current stream's list
+      if (p == nullptr && (sync_blocks_.find(stream) != sync_blocks_.end())) {
+          p = block_from_sync_list(stream);
+      }
+
+      // no blocks available, so create more
+      if (p == nullptr) {
+        new_blocks_from_heap(stream);
+        p = get_block(stream);
+      }
+    }
+
+    return p;
+  }
+
+  // Allocate new blocks from the heap memory resource into the free list
+  void new_blocks_from_heap(cudaStream_t stream) {
+    void* p = heap_resource_->allocate(heap_chunk_size_, stream);
+    if (!p) throw std::bad_alloc();
```

Comment:
```suggestion
```
This shouldn't be possible. If the upstream resource fails, it should throw it's own exception.